lms/djangoapps/completion/api/v1/views.py

```diff
@@ -0,0 +1,80 @@
+""" API v1 views. """
+import logging
+
+from django.contrib.auth.models import User
+from django.core.exceptions import ValidationError
+
+from rest_framework.views import APIView
+from rest_framework.response import Response
+from rest_framework import permissions
+
+from lms.djangoapps.completion.models import BlockCompletion
+from openedx.core.lib.api.permissions import IsStaffOrOwner
+
+log = logging.getLogger(__name__)
+
+
+class CompletionBatchView(APIView):
+    permission_classes = (permissions.IsAuthenticated, IsStaffOrOwner,)
+
+    def _validate(self, batch_object):
+        """
+        Performs validation on the batch object to make sure it is in the proper format.
+
+        Parameters:
+            batch_object: The object provided by the caller.
+
+        Raises:
+
+            KeyError:
+                If a key is not found in the object.
+
+            django.core.exceptions.ValidationError:
+                If a float is passed that is not between 0.0 and 1.0.
+        """
+        if 'user' not in batch_object:
+            raise KeyError("Key 'user' not found.")
+        if 'course_key' not in batch_object:
+            raise KeyError("Key 'course_key' not found.")
+        if 'blocks' not in batch_object:
+            raise KeyError("Key 'blocks' not found.")
+
+        return
```

Comment:
I'd like this method to do a full validation of the input data, and return a new dict containing the `User` object, a `CourseKey`, and a dict of `UsageKey` to `float`s which can be passed through to `submit_batch_completion`, so that `submit_batch_completion` doesn't have to do more validation.

FYI:
* `user = User.objects.get(username=user)` will throw `User. ObjectDoesNotExist` if no matching username is found.  But you can also use instead [`django.shortcuts.get_object_or_404`](https://github.com/edx/edx-platform/blob/6d2468d0dcf782a7561ecc61946146d6fae2762f/openedx/core/djangoapps/user_api/permissions.py#L30-L33)
* `course_key_obj = CourseKey.from_string(course_id)` will throw `opaque_keys.InvalidKeyError` if the `course_id` is not a valid course ID.
* Ditto for `UsageKey`

Reply:
Also, instead of raising `KeyErrors` here which get caught below, you can instead raise [django.http errors](https://docs.djangoproject.com/en/1.11/ref/request-response/#httpresponse-subclasses), which will return the status code and your error text to the user.

e.g see https://github.com/edx/edx-platform/blob/6d2468d0dcf782a7561ecc61946146d6fae2762f/openedx/core/djangoapps/user_api/permissions.py#L30-L33